db.go

```diff
@@ -482,6 +482,14 @@ func (db *DB) mmap(minsz int) error {
 
 // munmap unmaps the data file from memory.
 func (db *DB) munmap() error {
+	defer func() {
+		db.dataref = nil
```

Comment:
Updated (see below). I prefer not to have the suffix "`Unsafe`", and not to add the comment something like `it should be execute under the exclusive mmapLock and metaLock` as well.  BoltDB has well-designed lock mechanism (see below), we shouldn't worry about the lock on each single internal method/function. 

For write transaction, it requires the `rwlock` at [db.rwlock.Lock()](https://github.com/etcd-io/bbolt/blob/ad85400db2f3d67b391813f334a61e18b45c20c8/db.go#L732), and require the `mmaplock` when it needs to allocate more space accordingly needs to remap the db (of course release the lock when it finishes mapping), see [db.mmaplock.Lock()](https://github.com/etcd-io/bbolt/blob/ad85400db2f3d67b391813f334a61e18b45c20c8/db.go#L428).

For readonly transaction, it only requires the `mmaplock` at [db.mmaplock.RLock()](https://github.com/etcd-io/bbolt/blob/ad85400db2f3d67b391813f334a61e18b45c20c8/db.go#L695), to prevent the db file from being remapping by anther write transaction.

The new added `db.invalidateUnsafe()` is only called in `db.munmap()`, so it's safe.

```
func (db *DB) invalidate() {
	db.dataref = nil
	db.data = nil
	db.datasz = 0

	db.meta0 = nil
	db.meta1 = nil
}

```

Reply:
There is etcd convention (that I though is even golang convention, but I haven't found any signs) to call 
methods 'UnsafeFoo' when the caller of such method needs to take care of proper locking before calling such method. 

```
$ git grep Unsafe | grep func | nl
```
shows 93 cases. 

I do think that the call of 'invalidate' is safe in that context. I just want to avoid situation that someone will call 'invalidate' from context where the locks are not taken... because the person haven't read carefully that the body of the method interacts with the `db.data` and `db.meta` and these fields must be mutated under the exclusive lock. 

I'm not insisting on the "Unsafe" convention... but I don't think comment costs as here.

Reply:
Yes, etcd has the convention (`UnsafeFoo` vs `Foo`), but there is NO such convention in bbolt for now. If we change it to `invalidateUnsafe`, then it isn't consistent with other internal methods.

So I'd like to keep it as it's for now.

Of course, please feel free to raise a separate ticket if you do think we should follow the same convention as etcd.